python/ts/scriptqueue/script_queue.py

```diff
@@ -0,0 +1,287 @@
+        if not os.path.isdir(standardpath):
+            raise ValueError(f"No such dir standardpath={standardpath}")
+        if not os.path.isdir(externalpath):
+            raise ValueError(f"No such dir externalpath={externalpath}")
+
+        super().__init__(SALPY_ScriptQueue, 0)
+        self.cmd_stop.allow_multiple_callbacks = True
+        self.cmd_terminate.allow_multiple_callbacks = True
+        self.model = QueueModel(standardpath=standardpath, externalpath=externalpath,
+                                queue_callback=self.put_queue,
+                                script_callback=self.put_script)
+        self.put_queue()
+
+    def do_showAvailableScripts(self, id_data=None):
+        """Output a list of available scripts.
+
+        Parameters
+        ----------
+        id_data : `salobj.CommandIdData` (optional)
+            Command ID and data. Ignored.
+        """
+        self.assert_enabled("showAvailableScripts")
+        scripts = self.model.find_available_scripts()
+        evtdata = self.evt_availableScripts.DataType()
+        evtdata.standard = ":".join(scripts.standard)
+        evtdata.external = ":".join(scripts.external)
+        self.evt_availableScripts.put(evtdata, 1)
+
+    def do_showQueue(self, id_data):
+        """Output the queue event.
+
+        Parameters
+        ----------
+        id_data : `salobj.CommandIdData` (optional)
+            Command ID and data. Ignored.
+        """
+        self.assert_enabled("showQueue")
+        self.put_queue()
+
+    def do_showScript(self, id_data):
+        """Output the script event for one script.
+
+        Parameters
+        ----------
+        id_data : `salobj.CommandIdData` (optional)
+            Command ID and data. Ignored.
+        """
+        self.assert_enabled("showScript")
+        script_info = self.model.get_script_info(id_data.data.salIndex,
+                                                 search_history=True)
+        self.put_script(script_info)
+
+    def do_pause(self, id_data):
+        """Pause the queue. A no-op if already paused.
+
+        Unlike most commands, this can be issued in any state.
+
+        Parameters
+        ----------
+        id_data : `salobj.CommandIdData` (optional)
+            Command ID and data. Ignored.
+        """
+        self.model.running = False
+
+    def do_resume(self, id_data):
+        """Run the queue. A no-op if already running.
+
+        Parameters
+        ----------
+        id_data : `salobj.CommandIdData` (optional)
+            Command ID and data. Ignored.
+        """
+        self.assert_enabled("resume")
+        self.model.running = True
+
+    async def do_add(self, id_data):
+        """Add a script to the queue.
+
+        Start and configure a script SAL component, but don't run it.
+        """
+        self.assert_enabled("add")
+        script_info = ScriptInfo(
+            index=self.model.next_sal_index,
```

Comment:
I think it would be good if `do_add` returned the script index. This would give the user a direct way to check the state of a requested script.

Reply:
The only thing do_add can return to a remote user is the command ID. That is why the `script` event includes command ID -- it makes a fairly simple and completely reliable way to associate a particular script with a particular add command.

Reply:
Got it!